src/realm/object-store/impl/collection_notifier.cpp

```diff
@@ -132,12 +147,83 @@ DeepChangeChecker::DeepChangeChecker(TransactionChangeInfo const& info, Table co
 {
 }
 
+bool DeepChangeChecker::do_check_for_collection_modifications(std::unique_ptr<CollectionBase> coll, size_t depth)
+{
+    REALM_ASSERT(coll);
+    bool supported_collection_type = false;
+    if (auto lst = dynamic_cast<LnkLst*>(coll.get())) {
+        TableRef target = lst->get_target_table();
+        return std::any_of(lst->begin(), lst->end(), [&, this](auto key) {
+            return this->check_row(*target, key.value, depth + 1);
+        });
+    }
+    else if (auto dict = dynamic_cast<Dictionary*>(coll.get())) {
+        auto target = dict->get_target_table();
+        TableRef cached_linked_table;
+        for (auto it = dict->begin(); it != dict->end(); ++it) {
+            Mixed value = (*it).second;
+            if (value.is_type(type_TypedLink)) {
+                auto link = value.get_link();
+                REALM_ASSERT(link);
+                if (!cached_linked_table || cached_linked_table->get_key() != link.get_table_key()) {
+                    cached_linked_table = dict->get_table()->get_parent_group()->get_table(link.get_table_key());
+                    REALM_ASSERT_EX(cached_linked_table, link.get_table_key().value);
+                }
+                if (check_row(*cached_linked_table, link.get_obj_key().value, depth + 1)) {
+                    return true;
+                }
+            }
+            else if (value.is_type(type_Link)) {
+                REALM_ASSERT(target);
+                if (check_row(*target, value.get<ObjKey>().value, depth + 1)) {
+                    return true;
+                }
+            }
+        }
+        supported_collection_type = true;
+    }
+    else if (auto set = dynamic_cast<LnkSet*>(coll.get())) {
+        auto target = set->get_target_table();
+        REALM_ASSERT(target);
+        for (auto it = set->begin(); it != set->end(); ++it) {
+            ObjKey obj_key = *it;
+            if (obj_key && check_row(*target, obj_key.value)) {
```

Comment:
Can a `LnkSet` actually contain null even though `LnkLst` can't?

Reply:
Yes, LnkSet can contain a null if someone inserts a null ObjKey{}. Not sure if that discrepancy with LnkLst has been considered or not.